core/middlewares/with-routes.ts

```diff
@@ -282,21 +289,36 @@ export const withRoutes: MiddlewareFactory = () => {
     };
 
     if (route?.redirect) {
+      // Only carry over query params if the fromPath does not have any, as Bigcommerce 301 redirects support matching by specific query params.
+      const fromPathHasParams = new URL(route.redirect.fromPath, request.url).search.length > 0;
+      const searchParams = fromPathHasParams ? '' : request.nextUrl.search;
+
       switch (route.redirect.to.__typename) {
         case 'BlogPostRedirect':
         case 'BrandRedirect':
         case 'CategoryRedirect':
         case 'PageRedirect':
         case 'ProductRedirect': {
           // For dynamic redirects, assume an internal redirect and construct the URL from the path
-          const redirectUrl = new URL(route.redirect.to.path, request.url);
+          const redirectUrl = new URL(route.redirect.to.path + searchParams, request.url);
+
+          return NextResponse.redirect(redirectUrl, redirectConfig);
+        }
+
+        case 'ManualRedirect': {
+          // For manual redirects, to.url will be a relative path if it is an internal redirect and an absolute URL if it is an external redirect.
+          // URL constructor will correctly handle both cases.
+          // If the manual redirect is an external URL, we should not carry query params.
+          const isExternalRedirect =
+            new URL(route.redirect.to.url, request.url).origin !== request.nextUrl.origin;
+          const manualSearchParams = isExternalRedirect ? '' : searchParams;
+          const redirectUrl = new URL(route.redirect.to.url + manualSearchParams, request.url);
```

Comment:
Similar thing here, but just to make it more concise.
```suggestion
          const redirectUrl = new URL(route.redirect.to.url, request.url);

          if (redirectUrl.origin === request.nextUrl.origin) {
            redirectUrl.search = searchParams;
          }
```